### ISSUE TITLE
Permutation Argument Soundness

### ISSUE THREAD

**Issue by Bri-K** [aszepieniec/stark-brainfuck#43]
The argument in point 1. of the section "Permutation Arguments" appears to be incomplete.

Denote the tables as matrices by A, B and the verifier randomness (a, b, c, ...) by the vector r. What needs to be analyzed is the probability that the components of the vector Ar are a permutation of the components of the vector Br when the rows of A and B are _not_ permutations of each other. That is, what are the odds that for every permutation matrix P, A ≠ PB, but that we sample an r such that Ar = PBr for some P?

The argument given correctly analyzes the probability of sampling an r such that Ar = PBr for a fixed P. However, if Ar = PBr for _any_ permutation matrix P, then the prover can trick the verifier, so only considering a single P is insufficient.

Basically, it's _a priori_ possible that A - PB has a large codimension 1 kernel for a large number of permutations P. If these kernels are distinct they could fill up a significant portion of the space the verifier is sampling from, since the number of permutations P is on par with the field size for tables of even moderate height.

**Comment by aszepieniec** [comment]
Excellent remark. I agree with this analysis entirely. Well done!

Here is an updated argument that works around the issue:

Let b be a row of B that does not exist in A. The malicious prover wins only if he receives challenges r such that for some row a of A, ar = br. The probability of this event for a given row a is 1/|F|. By the union bound, the probability of this event for any row a of A is at most N/|F|, where N is the number of rows.
